### PR TITLE
Problem: disallow same sender tx appear in the same tx after contract

### PR DESCRIPTION
**Pull request by yihuang**
creation

Closes: #XXX

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: app/ante/eth.go [hunk 1]
```diff
@@ -311,3 +311,38 @@ func CheckAndSetEthSenderNonce(
 
 	return nil
 }
+
+// DetectContractCreationBatchTx returns error if same same transaction appear after contract creation tx in the same
```
Comment (yihuang) [inline app/ante/eth.go]:
```suggestion
// DetectContractCreationBatchTx returns error if same sender transaction appear after contract creation tx in the same
```

**Comment by mmsqe** [comment]
might need adjust the test_batch_tx now

File: app/ante/eth.go [hunk 2]
```diff
@@ -311,3 +311,38 @@ func CheckAndSetEthSenderNonce(
...
+
+		sender := string(msgEthTx.GetSender().Bytes())
+		if _, ok := senders[sender]; ok {
+			return errorsmod.Wrapf(
+				errortypes.ErrInvalidRequest,
+				"same sender included after contract creation in the batch transaction",
+			)
+		}
+
+		if msgEthTx.AsTransaction().To() == nil {
```
Comment (mmsqe) [inline app/ante/eth.go:342]: do we need check if nonce is valid if transfer tx is in same batch of create contract?

Comment (yihuang) [reply app/ante/eth.go:342]: that's already checked by the normal nonce logic, right?

File: tests/integration_tests/hardhat/contracts/TestERC20A.sol [hunk 3]
```diff
@@ -4,7 +4,7 @@ pragma solidity 0.8.10;
 import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 
 contract TestERC20A is ERC20 {
-    constructor() public ERC20("TestERC20", "Test") {
-        _mint(msg.sender, 100000000000000000000000000);
+    constructor(owner address) public ERC20("TestERC20", "Test") {
```
Comment (mmsqe) [inline tests/integration_tests/hardhat/contracts/TestERC20A.sol]:
```suggestion
    constructor(address owner) public ERC20("TestERC20", "Test") {
```

**Review by mmsqe — APPROVED** [review]

**Pull request merged**